From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Bhadram Varka <vbhadram@nvidia.com>
Date: Fri, 18 Nov 2022 11:31:45 +0530
Subject: [PATCH 037/243] NVIDIA: SAUCE: net: phy: aqr113c: skip PHY low power
 mode

Issue: Uncorrectable Errors are observed when the MGBE1
the interface is going down or in SC7 since PHY is put into
low power mode without powering the UPHY lanes and
UPHY lanes can't be power down in ethernet server enabled
case where lanes are shared with other VF's.

Fix: Skip PHY low power mode in automotive platforms.

Bug 3683205

Change-Id: I09c700165f85856cff01838184847eaf581615f3
Signed-off-by: Bhadram Varka <vbhadram@nvidia.com>
Signed-off-by: Revanth Kumar Uppala <ruppala@nvidia.com>
Reviewed-on: https://git-master.nvidia.com/r/c/linux-stable/+/2909754
Tested-by: Abhilash G <abhilashg@nvidia.com>
Reviewed-by: Abhilash G <abhilashg@nvidia.com>
Reviewed-by: Laxman Dewangan <ldewangan@nvidia.com>

diff --git a/drivers/net/phy/aquantia/aquantia.h b/drivers/net/phy/aquantia/aquantia.h
index 89655b6e4004..666755125df6 100644
--- a/drivers/net/phy/aquantia/aquantia.h
+++ b/drivers/net/phy/aquantia/aquantia.h
@@ -231,6 +231,7 @@ struct aqr107_priv {
 	bool wait_on_global_cfg;
 	struct aqr_global_syscfg global_cfg[AQR_NUM_GLOBAL_CFG];
 	int wol_status;
+	bool skip_lpm;	/* skip low power mode */
 };
 
 #if IS_REACHABLE(CONFIG_HWMON)
diff --git a/drivers/net/phy/aquantia/aquantia_main.c b/drivers/net/phy/aquantia/aquantia_main.c
index 1143efd50a91..d492d5b99f9a 100644
--- a/drivers/net/phy/aquantia/aquantia_main.c
+++ b/drivers/net/phy/aquantia/aquantia_main.c
@@ -1223,8 +1223,12 @@ static int aqr_gen2_get_rate_matching(struct phy_device *phydev,
 
 static int aqr_gen1_suspend(struct phy_device *phydev)
 {
+	struct aqr107_priv *priv = phydev->priv;
 	int err;
 
+	if (priv->skip_lpm)
+		return 0;
+
 	err = phy_set_bits_mmd(phydev, MDIO_MMD_VEND1, MDIO_CTRL1,
 			       MDIO_CTRL1_LPOWER);
 	if (err)
@@ -1235,8 +1239,12 @@ static int aqr_gen1_suspend(struct phy_device *phydev)
 
 static int aqr_gen1_resume(struct phy_device *phydev)
 {
+	struct aqr107_priv *priv = phydev->priv;
 	int err;
 
+	if (priv->skip_lpm)
+		return 0;
+
 	err = phy_clear_bits_mmd(phydev, MDIO_MMD_VEND1, MDIO_CTRL1,
 				 MDIO_CTRL1_LPOWER);
 	if (err)
@@ -1339,6 +1347,8 @@ static int aqr_gen2_config_inband(struct phy_device *phydev, unsigned int modes)
 
 static int aqr107_probe(struct phy_device *phydev)
 {
+	struct device_node *node = phydev->mdio.dev.of_node;
+	struct aqr107_priv *priv;
 	int ret;
 
 	phydev->priv = devm_kzalloc(&phydev->mdio.dev,
@@ -1350,6 +1360,12 @@ static int aqr107_probe(struct phy_device *phydev)
 	if (ret)
 		return ret;
 
+	priv = phydev->priv;
+	if (of_property_read_bool(node, "aquantia,skip-lpm"))
+		priv->skip_lpm = true;
+	else
+		priv->skip_lpm = false;
+
 	return aqr_hwmon_probe(phydev);
 }
 
-- 
2.43.0

